README.md
# Roadroller: Flattens Your JavaScript Demo

**Roadroller** is a heavyweight JavaScript packer for large [demos][demo]. It was originally designed for [js13kGames], but it remains usable for demos as small as 4KB. Depending on the input it can provide up to 15% additional compression compared to best ZIP/gzip recompressors. **[Try it online][online]!**

Roadroller is considered "heavyweight" unlike typical JS packers such as [JSCrush] or [RegPack], because it is quite resource intensive and requires both a considerable amount of memory and a non-negligible run time. The default should work for most devices, but you can configure both aspects as you need.

## Quick Start

In addition to the [online demo][online], Roadroller is available as an [NPM package][npm]:

```
$ npx roadroller input.js -o output.js
```

You can also use Roadroller as a library to integrate with your build pipeline.

```javascript
import { Packer } from 'roadroller';

const inputs = [
    {
        data: 'console.log("Hello, world!");',
        type: 'js',
        action: 'eval',
    },
];

const options = {
    // see the Usage for available options.
};

const packer = new Packer(inputs, options);
await packer.optimize(); // takes less than 10 seconds by default

const { firstLine, secondLine } = packer.makeDecoder();
console.log(firstLine + secondLine);
```

Roadroller as a library or a CLI command requires Node.js 14 or later. Node.js 16 is strongly recommended because Roadroller is substantially faster in 16 than in 14.

## Usage

By default Roadroller receives your JS code and returns a compressed JS code that should be further compressed with ZIP, gzip or PNG bootstrap (or more accurately, [DEFLATE]). Ideally your JS code should be already minified, probably using [Terser] or [Closure Compiler]; Roadroller only does a minimal whitespace and comment suppression.

The resulting code will look like this: (the newline is mostly for the explanation and can be removed)

```javascript
eval(Function("[M='Zos~ZyF_sTdvfgJ^bIq_wJWLGSIz}Chb?rMch}...'"
,...']charCodeAtUinyxp',"for(;e<12345;c[e++]=p-128)/* omitted */;return o")([],[],12345678,/* omitted */))
```

The first line is a compressed data. It can contain control characters like `` (U+001C) that might not render in certain environments. Nevertheless you should make sure that they are all copied in verbatim.

The second line is a compressor tuned for this particular input. By default the decompressed data immediately goes through `eval`, but you can configure what to do with that.

The first line is very incompressible unlike the second line, so ideally you should compress two lines separately. This is best done by using ADVZIP from [AdvanceCOMP] or [ECT]. The first line and second line may form a single statement as above so they should not be separated; you can only put whitespace between them.

<!--
### Multiple Inputs

You can also give multiple inputs to Roadroller; for example you can put shaders and map data. The executed code will receive all decompressed inputs. This is more efficient than putting them into strings because each input can be separately modelled.
-->

### Input Configuration

Each input can be further configured by input type and action. In the CLI you put corresponding options *before* the file path.

**Input type** (CLI `-t|--type TYPE`, API `type` in the input object) determines the preprocessing step to improve the compression. <!--Dropping a file to the input window also tries to detect the correct input type.-->

* **JavaScript** (`js`) assumes a valid JS code. Automatically removes all redundant whitespace and comments and enables a separate modelling for embedded strings. This also works for JSON.

<!--* **GLSL** (`glsl`) assumes a valid GLSL code. Automatically removes all redundant whitespace and comments.-->

<!--* **HTML** (`html`) assumes a valid HTML. (TODO)-->

* **Text** (`text`) assumes a human-readable Unicode text that can be encoded in UTF-8. This can also be used for JavaScript code that should not undergo preprocessing.

<!--* **Binary** (`binary`) does nothing. You can choose base64 (`binary:base64`) or hex (`binary:hex`) for the input encoding.-->

**Input action** (CLI `-a|--action ACTION`, API `action` in the input object) determines what to do with the decompressed data. <!--All action except for the evaluate produces a value to the variable named `_` by default, which is either a value itself for a single input and an array of values for multiple inputs.-->

* <!--*(JS, text only)*-->
  **Evaluate** (`eval`) evaluates the decompressed JavaScript code. If there are multiple inputs there should be exactly one JavaScript input with evaluate action, since subsequent inputs will be decompressed in that code. The resulting value is always a code string, which may include decoders for subsequent inputs.

<!--* *(JS, text only)* **JSON decode** (`json`) parses and returns a JSON value with `JSON.parse`.-->

<!--* *(No binary)* **String** (`string`) returns a string.-->

* <!--*(No binary)*-->
  **Write to document** (`write`) writes a decompressed string to `document`. Typically used with HTML.

<!--* **Array** (`array`) returns an array of bytes.-->

<!--* **Typed array** (`u8array`) returns a `Uint8Array` value.-->

<!--* **Base64** (`base64`) returns a base64-encoded string. Handy for data URIs.-->

<!--
**Input name** (CLI `-n|--name NAME`, API `name`) is required for accessing each input from the decompressed code. This is required if the input produces an output value.

**Extract inputs** (CLI `-x|--extract`, not available in API) can be used for the JavaScript input with the evaluate action. This will try to extract long embedded strings and determine the best type for each input. This assumes that the compressed code doesn't make use of the output variable elsewhere; you can change the variable name from the configuration.
-->

### Output Configuration

**Number of contexts** (CLI `-S|--selectors xCOUNT`) relates to the complexity of modelling. The larger number of contexts will compress better, but at the expense of linear increase in both the time and memory usage. The default is 12, which targets at most 1 second of latency permitted for typical 30 KB input.

**Maximum memory usage** (CLI `-M|--max-memory MEGABYTES`, API `maxMemoryMB` in the options object) configures the maximum memory to be used for decompression. Increasing or decreasing memory usage mostly affects the compression ratio and not the run time. The default is 150 MB and a larger value is not recommended for various reasons:

* Any further gain for larger memory use is negligible for typical inputs less than 100 KB.

* The compression may use more memory than the decompression: an one-shot compression may use up to 50% more memory, the optimizer will use 50% more on top of that.

* It does take time to allocate and initialize a larger memory (~500 ms for 1 GB), so it is not a good choice for small inputs.

The actual memory usage can be as low as a half of the specified due to the internal architecture; `-v` will print the actual memory usage to stderr.

**Allowing the decoder to pollute the global scope** (CLI `-D|--dirty`, API `allowFreeVars` in the options object) is unsafe especially when the Roadroller output should coexist with other code or there are elements with single letter `id` attributes and turned off by default. But if you can control your environment (typical for demos), you can turn this on for a smaller decoder.

**Optimize parameters** (CLI `-O|--optimize LEVEL`, API `Packer.optimize`) searches for better modelling parameters. If parameters are already given the optimizer will try to improve upon that, and the optimizer prints best parameters at the end which can be reused for faster iteration. Parameters are solely related to the compression ratio so you can try this as many as you can afford. Each level does the following:

* Level 0 does absolutely nothing and uses given parameters or default parameters if none. This is the default when any optimizable parameters are given.

* Level 1 runs a quick search with about 30 sets of parameters and takes less than 10 seconds for typical 30 KB input. This is the default when no optimizable parameters are given, and intended for the typical build process.

* Level 2 runs a thorough search with about 300 sets of parameters and takes about a minute or two. This is best useful for the release build and you would like to save best parameters for later uses.

* Level ∞ is a special option only available in the CLI (`-OO`, with two capital Ohs) and runs increasingly slower optimizations in a run. Once the highest level is reached it runs that level forever. You need to explicitly terminate the search (e.g. CTRL-C), then it will proceed with the best parameters so far.

### Advanced Configuration

**Number of context bits** (CLI `-Zco|--context-bits BITS`, API `contextBits` in the options object) sets the size of individual model as opposed to the total memory use (`-M`), which is a product of the number of context and the size of each model. This explicit option is most useful for the fair benchmarking, since some parameters like `-Zpr` or `-Zmc` affect the memory use and therefore this parameter.

<!--**Optimize for uncompressed size** (CLI `--uncompressed-only`) assumes the absence of the outer comperssion algorithm like DEFLATE. This is *bad* for the compression since the compressor has to work strictly within the limits of JS source code including escape sequences. This should be the last resort where you can't even use the PNG-based self extraction and everything has to be in a single file.-->

> Following parameters can be automatically optimized and normally you don't have to touch them unless you want to reproduce a particular set of parameters. As such, the default optimization (`-O1`) is disabled if any of these arguments are given in the CLI.


CHANGELOG.md
# Changelog

## v2.1.0 (2021-09-13)

Compression improvements:

- The compressor now uses a specialized WebAssembly implementation by default. (c2bd4353b1d22407ced62353fe5acffa0ce0487b, d9a38973cece36427e16671fc83ff98d91e0428c)
- Cache the result of JavaScript code transformation during the optimization. Combined with the WebAssembly implementation, the optimization is up to 2.5 times faster than 2.0.0. (02d09a3906a443c6ee6f7edff310a2ac17b06e16)
- Removed a special result by the optimizer for turning JS inputs into texts and made the equivalent a separate option (`-Zdy`). This also allows for using JS-specific modelling for texts if beneficial, and avoids the UI confusion present in the online demo. (4dd5fe7261023b170b927811c4b1a86c05432a4f)
- The rANS state is now embedded directly into the compressed data. (48ace3df9411840b2417b8ccbe43a09bf9d9e5dc)
- The decoder is now slightly smaller. (819a574c21b4d220f1290299c698c34b25e8bd98, eb746302f8a598bea733111b9ea964882a8f3393, c6a88ee2764f9d00b47561cf35250c63b58ddf27)
- Tools no longer generate a newline between the first and second "lines". (eb06eddcb5c10f6f39d232b3dadce3decbb4295e)

UI/API improvements:

- `-OO` option is added to the CLI, which runs a series of optimizations forever until a user request. The optimization now proceeds using the best parameters so far when the user has aborted the search with Ctrl-C even in other optimization levels. (974a9ee1204011f5b76852b4474938e2bf50c8e2)
- Renamed `ArrayBufferPool` to `ResourcePool` to allow using it for other uses (e.g. WebAssembly modules). Also made it clear that `ResourcePool.allocate` should not be used outside of Roadroller. (3f073cf451abae9875471dbe1e4915723393162b)
- Packer now creates its own `ResourcePool` if not given, so `Packer.optimize` is now equally fast with or without the pool. (2c727ea5c4aa77ec49b27c3f222be424551fa7de)
- The online demo deallocates the pool a minute after the last compression. (5e86a59e80441db458ff7c4e114d7248d21e5c59)
- `compressWithDefaultModel` has been added as a shortcut (and optimized version) for `compressWithModel` with `DefaultModel`. (1a644e95422b86352eb376605bfa2a8852013ddf)

Bugfixes:

- Fixed a typo in `index.d.ts`. (3f073cf451abae9875471dbe1e4915723393162b)

Deprecations:

- `ArrayBufferPool` and `arrayBufferPool` fields are now deprecated in favor of `ResourcePool` and `resourcePool`. (3f073cf451abae9875471dbe1e4915723393162b)
- `OptimizedPackerOptions.preferTextOverJS` is no longer used and thus deprecated. (14292282e2b926862ca580fde9bca125d559874f)

## v2.0.0 (2021-08-28)

**Breaking Changes:**

- **The first line and second line are no longer guaranteed to be separate statements.** In particular the documentation accidentally suggested that two lines can be (but are not recommended to be) in separate files, necessitating a breaking change.
- `DirectContextModelOptions.modelRecipBaseCount` is now required. The value of 2 is identical to the 1.x behavior.
- Deprecated APIs have been removed.
    - `OptimizerProgressInfo.temperature`: Use `ratio` instead.
    - `optimizeSparseSelectors` function and `Packer.optimizeSparseSelectors` method: Use `Packer.optimize` instad.
    - `PackerOptions.learningRateNum/Denom`: Use `PackerOptions.recipLearningRate` instead.

Compression improvements:

- Added a new tunable `modelRecipBaseCount` parameter (`-Zmd` in the CLI) with a new default, allowing ~3% additional improvements. (4e8041c11457636b6a171e316752af92d4d79614, 207186191cfdb6dfc1251f597b39284ae20dd89c, 717f7eeb294d55c72fb53f5ae31bdbebbf8c59a5)
- The decoder is now slightly smaller. (9cf184518c93005a386895f381ba1ffc6e1b47e3)

UI/API improvements:

- #14: `-O1` is now a quick search that only takes ~10s for typical inputs. The old `-O1` is renamed to `-O2`. (df5c8edb6a803f47dfb3d29c1dc6d5fcba738322)
- The CLI now defaults to `-O1` if no optimizable arguments are given. Use `-O0` to explicitly disable the optimization. (a610f680a3d1e5aabed81368af050ac5271b2a55)
- The estimated compressed size is now more accurate. (8732d3da350b8da58e843921d36fa509fe0737ae)
- The online demo is now slightly more mobile friendly. (10b0bbe6715e36175421c4cac5360148c30fe942)

Bugfixes:

- #7: The output no longer pollutes the global scope by default. New `allowFreeVars` option (`-D` in the CLI) can be used to allow the decoder to be "dirty". (70fae12ee0eeaf754d46e6db78dacd1097f5c5cb, ecd8116c2c0c305334e18fea99b25d9f72b82d5e)
- #15: The optimizer now works in Node.js 14. It is still strongly recommended to use Node.js 16 because 14 is significantly slower. (13d9c8d39740790b5bfcc3b07035a41510aba19f)
- Fixed a bug where `-Zco -M` (but not `-M -Zco`) was incorrectly allowed in the CLI. (ecd8116c2c0c305334e18fea99b25d9f72b82d5e)
- `Packed.freeVars` was missing one variable. (740f2b5eae65b0c033c5f382f49742393d02118a)

## v1.2.1 (2021-08-24)

Compression improvements:


cli.mjs
#!/usr/bin/env node

import * as fs from 'fs';
import process from 'process';
import * as readline from 'readline';
import { performance } from 'perf_hooks';
import { ResourcePool, Packer, defaultSparseSelectors } from './index.mjs';

let VERSION = 'unknown';
try {
    VERSION = JSON.parse(fs.readFileSync(new URL('./package.json', import.meta.url), { encoding: 'utf-8' })).version;
} catch (e) {
}

function usage(verbose) {
    console.warn(`\
     ____                                                     ___
     /    )                   /              /  /            |   |
    /___ /    __    __    __ /  )__    __   /  /   __   )__  |_JS|
   /    |   /   ) /   ) /   /  /   ) /   ) /  /  /___) /   )
__/_____|__(___/_(___(_(___/__/_____(___/_/__/__(___ _/______${VERSION}__

Usage: roadroller [OUTPUTOPTIONS] [INPUTOPTIONS] [--] INPUT ...

Input options: (only affects the next input path)
INPUT
  The input path. Can be given multiple times but only one is supported
  at the moment. Can be "-" for stdin.
--
  Can be followed by input options to signal that the next argument is
  always an input path even if it starts with "-".
  NOTE: If there are multiple input paths starting with "-"
  there should be multiple "--" for each input.
-t|--type TYPE
  Sets the input type. Inferred from the file extension if omitted.
    js              Valid JavaScript code (UTF-8) [Default for .js]
    text            Human-readable text (UTF-8)
-a|--action ACTION
  Sets the input action. Inferred from the input type if omitted.
    eval            Evaluates as JavaScript code [Default for -t js]
    write           Writes to document

Output options:
-o|--output-file OUTPUT [Default: -]
  The output path. Can be "-" for stdout.
-O|--optimize EFFORTS [Default: 0 or 1]
  Tries to tune parameters for this input.
    0   Use the baseline parameters.
        Default when any optimizable arguments are given.
    1   Tries to optimize -S and most -Z arguments with ~30 attempts.
        Default when no optimizable arguments are given.
    2   Same to -O1 but with ~300 attempts.
    O (capital letter Oh, looks like the infinity symbol ∞)
        Continuously runs -O1 through the maximal level, then
        repeatedly runs the maximal level until the user aborted.
  Anything beyond -O0 prints the best parameters unless -q is given.
  Pressing Ctrl-C (SIGINT) anytime aborts the search and proceeds
  with the best parameters so far.
-M|--max-memory MEGABYTES [Range: 10..1024, Default: 150]
  Configures the maximum memory usage.
  The actual usage might be lower. Use -v to print the actual usage.
-D|--dirty [Default: false]
  Allow the decoder to pollute the global scope.
  This is unsafe in general, but if you know that this is safe
  (i.e. no script in the global scope, no single-letter ids in use)
  then this can shave a few more bytes from the decoder.
-S|--selectors xNUMCONTEXTS [Range: 1..64, Default: 12]
  Sets the maximum number of contexts used, prefixed by a literal "x".
  For 13+ contexts additional contexts are randomly picked.
  Can be used for finer (linear) tuning of the memory usage.
` + (verbose > 0 ?
`-S|--selectors SELECTOR,SELECTOR... [Default: ${defaultSparseSelectors()}]
  Sets the explicit contexts to be used.
  Each number designates bytes to be used as a context for that model;
  if K-th (K>0) lowest bit is set the context uses the K-th-to-last byte,
  so for example 5 = 101 (binary) designates a (0,2) sparse context model.
-Zab|--num-abbreviations NUM [Range: 0..64, Default: 64]
  Limit the number of name abbreviations to NUM.
  Ignored if there are less than NUM feasible abbreviations.
-Zco|--context-bits BITS [Range: 1..24+, Default: derived]
  Sets the size of each context model, as opposed to the total size (-M).
  The maximum can range from 24 to 30 depending on the number of contexts.
-Zdy|--dynamic-models FLAGS [Default: 1 for JS, 0 for others]
  Enables or disables specific dynamic models.
  The value is a bitwise OR of the following bits:
    1        Puts quoted strings (', ", \`) into a separate context.
             This is only useful when every quotes are paired, so it
             can't be used in English texts with contractions (e.g. isn't).
-Zlr|--learning-rate RATE [Range: 1..2^53, Default: 500]
  Configures the learning rate of context mixer; smaller adapts faster.
-Zmc|--model-max-count COUNT [Range: 1..32767, Default: 5]
  Configures the adaptation speed of context models.
  Context models adapt fastest when the context is first seen,
  but become slower each subsequent occurrence of the context.
  This option configures how slowest the adaptation can be;
  smaller COUNT is better for quickly varying (non-stationary) inputs.
-Zmd|--model-base-divisor COUNT [Range: 1..2^53-1, Default: 20]
  Configures the reciprocal of initial count when a context is first seen.
  The count is increased by 1 per occurrence and larger count decreases
  the speed of adaptation, so if this option is larger the initial count
  is made smaller and the model would behave much quicker initially.
-Zpr|--precision BITS [Range: 1..21, Default: 16]
  Sets the precision of internal fixed point representations.
` : '') + `
Other options:
-q|--silent
  Suppresses any diagnostic messages.
-v|--verbose
  Prints more information whenever appropriate. Can be repeated.
-h|--help
  Prints this message. Combine with -v for more options.
-V|--version
  Prints the version.

Please see https://github.com/lifthrasiir/roadroller/ for more information.`);
}

function readToString(stream) {
    const chunks = [];
    return new Promise((resolve, reject) => {
        stream.on('data', (chunk) => chunks.push(Buffer.from(chunk)));
        stream.on('error', (err) => reject(err));
        stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf-8')));
    })
}

async function parseArgs(args) {
    const inputs = [];
    let currentInput = {};
    const options = {
        resourcePool: new ResourcePool(),
    };
    let command;
    let optimize;
    let defaultOptimize = 1;
    let outputPath;
    let nextIsArg = false;
    let verbose = 0; // -1: -q, 0: default, 1: -v, 2: -vv, ...

    while (args.length > 0) {
        const opt = args.shift();
        if (nextIsArg || opt === '-' || !opt.startsWith('-')) {
            nextIsArg = false;
            try {
                if (opt === '-') {
                    currentInput.data = await readToString(process.stdin);
                } else {
                    currentInput.data = fs.readFileSync(opt, { encoding: 'utf-8' });
                }
            } catch (e) {
                throw `cannot read from the input path ${opt}`;
            }
            if (currentInput.type === undefined) {
                if (opt.endsWith('.js')) {
                    currentInput.type = 'js';
                } else {
                    throw `cannot infer the input type for ${opt}`;
                }
            }
            if (currentInput.action === undefined) {
                if (currentInput.type === 'js') {
                    currentInput.action = 'eval';
                } else {
                    throw `cannot infer the input action for ${opt}`;
                }
            }
            inputs.push(currentInput);
            currentInput = {};
            continue;
        }

        const matchOpt = (long, short) =>
            opt === '--' + long || (short && opt === '-' + short);
        const matchOptArg = (long, short) =>
            opt.match(new RegExp(`^(?:--${long}|(?:--${long}=${short ? `|-${short}=?` : ''})(.*))$`));
        const getArg = m => {
            if (m[1]) return m[1];
            if (args.length === 0) throw `invalid ${opt} argument`;
            return args.shift();
        };

        let m;
        if (matchOpt('help', 'h')) {
            command = 'usage';
        } else if (matchOpt('version', 'V')) {
            command = 'version';
        } else if (matchOpt('silent', 'q')) {
            if (verbose > 0) throw '-q and -v cannot be used together';
            verbose = -1;
        } else if (m = opt.match(/^(?:--verbose|-(v+))$/)) {
            if (verbose < 0) throw '-v and -q cannot be used together';
            verbose += (m[1] || 'v').length;
        } else if (m = matchOptArg('type', 't')) {
            if (currentInput.type !== undefined) throw 'duplicate --type arguments';
            currentInput.type = getArg(m);
        } else if (m = matchOptArg('action', 'a')) {
            if (currentInput.action !== undefined) throw 'duplicate --action arguments';
            currentInput.action = getArg(m);
        } else if (m = matchOptArg('output-file', 'o')) {
            if (outputPath !== undefined) throw 'duplicate --output-file arguments';
            outputPath = getArg(m);
        } else if (m = matchOptArg('optimize', 'O')) {
            if (optimize !== undefined) throw 'duplicate --optimize arguments';
            const arg = getArg(m);
            optimize = arg === 'O' ? Infinity : parseInt(arg, 10);
        } else if (m = matchOptArg('max-memory', 'M')) {
            if (options.maxMemoryMB !== undefined) throw 'duplicate --max-memory arguments';
            options.maxMemoryMB = parseInt(getArg(m), 10);
        } else if (m = matchOptArg('dirty', 'D')) {
            if (options.allowFreeVars !== undefined) throw 'duplicate --dirty arguments';
            options.allowFreeVars = true;
        } else if (m = matchOptArg('selectors', 'S')) {
            if (options.sparseSelectors !== undefined) throw 'duplicate --max-memory arguments';
            const arg = getArg(m);
            let selectors;
            if (arg.match(/^x[0-9]+$/)) {
                const numContexts = parseInt(arg.substr(1), 10);
                if (numContexts < 1) throw 'no selectors in --selectors argument';
                if (numContexts > 64) throw 'too many selectors in --selectors argument';
                selectors = defaultSparseSelectors(numContexts);
            } else {
                selectors = arg.split(/,/g).map(v => parseInt(v, 10)).sort((a, b) => a - b);
                defaultOptimize = 0;
            }
            options.sparseSelectors = selectors;
        } else if (m = matchOptArg('num-abbreviations', 'Zab')) {
            if (options.numAbbreviations !== undefined) throw 'duplicate --num-abbreviations arguments';
            options.numAbbreviations = parseInt(getArg(m), 10);
            defaultOptimize = 0;
        } else if (m = matchOptArg('context-bits', 'Zco')) {
            if (options.contextBits !== undefined) throw 'duplicate --context-bits arguments';
            options.contextBits = parseInt(getArg(m), 10);
            // -Zco is not optimizable, so its use doesn't change -O defaults
        } else if (m = matchOptArg('dynamic-models', 'Zdy')) {
            if (options.dynamicModels !== undefined) throw 'duplicate --dynamic-models arguments';
            options.dynamicModels = parseInt(getArg(m), 10);
        } else if (m = matchOptArg('learning-rate', 'Zlr')) {
            if (options.recipLearningRate !== undefined) throw 'duplicate --learning-rate arguments';
            options.recipLearningRate = parseInt(getArg(m), 10);
            defaultOptimize = 0;
        } else if (m = matchOptArg('model-max-count', 'Zmc')) {
            if (options.modelMaxCount !== undefined) throw 'duplicate --model-max-count arguments';
            options.modelMaxCount = parseInt(getArg(m), 10);
            defaultOptimize = 0;
        } else if (m = matchOptArg('model-base-divisor', 'Zmd')) {
            if (options.modelRecipBaseCount !== undefined) throw 'duplicate --model-base-divisor arguments';
            options.modelRecipBaseCount = parseInt(getArg(m), 10);
            defaultOptimize = 0;
        } else if (m = matchOptArg('precision', 'Zpr')) {
            if (options.precision !== undefined) throw 'duplicate --precision arguments';
            options.precision = parseInt(getArg(m), 10);
            defaultOptimize = 0;
        } else if (opt == '--') {
            nextIsArg = true;
        } else {
            throw `unknown option ${opt}`;
        }
    }

    if (Object.keys(currentInput).length > 0) {
        throw 'the last input path is missing';
    }

    const between = (a, v, b) => a <= v && v <= b; // also returns false if v is NaN

    // check validity of arguments
    if (outputPath === undefined) {
        outputPath = '-';
    }
    if (optimize === undefined) {
        optimize = defaultOptimize;
    } else if (!between(0, optimize, 2) && optimize !== Infinity) {
        throw 'invalid --optimize argument';
    }
    if (options.maxMemoryMB !== undefined) {
        if (options.contextBits !== undefined) throw '--max-memory and --context-bits cannot be used together';
        if (!between(10, options.maxMemoryMB, 1024)) throw 'invalid --max-memory argument';
    } else if (options.contextBits !== undefined) {
        if (!between(1, options.contextBits, 30)) throw 'invalid --context-bits argument';
        const numSelectors = options.sparseSelectors ? options.sparseSelectors.length : 12;
        // this implies that -Zco24 unconditionally works but -Zco25..30 may not work depending on -Sx
        if (numSelectors * (1 << options.contextBits) > (1 << 30)) throw 'invalid --context-bits argument';
    }
    if (options.sparseSelectors !== undefined) {
        if (options.sparseSelectors.length < 1) throw 'no selectors in --selectors argument';
        if (options.sparseSelectors.length > 64) throw 'too many selectors in --selectors argument';
        if (options.sparseSelectors.some(sel => sel !== sel || sel < 0 || sel >= 0x80000000)) {
            throw 'invalid selector in --selectors argument';
        }
        if (options.sparseSelectors.find((sel, i) => options.sparseSelectors[i-1] === sel)) {
            throw 'duplicate selector in --selectors argument';
        }
    }
    if (options.numAbbreviations !== undefined && !between(0, options.numAbbreviations, 64)) {
        throw 'invalid --num-abbreviations argument';
    }
    if (options.dynamicModels !== undefined && !between(0, options.dynamicModels, 1)) {
        throw 'invalid --dynamic-models argument';
    }
    if (options.recipLearningRate !== undefined && !between(1, options.recipLearningRate, 2**53)) {
        throw 'invalid --learning-rate argument';
    }
    if (options.modelMaxCount !== undefined && !between(1, options.modelMaxCount, 32767)) {
        throw 'invalid --model-max-count argument';
    }
    if (options.modelRecipBaseCount !== undefined && !between(1, options.modelRecipBaseCount, 2**53 - 1)) {
        throw 'invalid --model-base-divisor argument';
    }
    if (options.precision !== undefined && !between(1, options.precision, 28)) {
        throw 'invalid --precision argument';
    }

    if (!command && inputs.length === 0) {
        command = 'default';
    }
    if (command) {
        return { command, verbose };
    }

    return { command: 'compress', inputs, options, optimize, outputPath, verbose };
}

async function compress({ inputs, options, optimize, outputPath, verbose }) {
    let packer = new Packer(inputs, options);
    const origLength = inputs.reduce((acc, { data } ) => {
        return acc + (Array.isArray(data) ? data.length : unescape(encodeURIComponent(data)).length);
    }, 0);

    if (verbose >= 1) {
        console.warn(
            `Actual memory usage: ${packer.memoryUsageMB < 1 ? '< 1' : packer.memoryUsageMB.toFixed(1)} MB` +
            (options.contextBits ? '' : ` (out of ${options.maxMemoryMB || 150} MB)`));
    }

    if (optimize) {
        const defaultSelectorsJSON =
            !options.sparseSelectors ?
                JSON.stringify(defaultSparseSelectors()) :
            options.sparseSelectors.length <= 12 ?
                JSON.stringify(defaultSparseSelectors(options.sparseSelectors.length)) :
                ''; // more than 13 selectors are randomly determined

        const format = moreOptions => {
            const combined = { ...options, ...moreOptions };
            let args;
            if (!combined.sparseSelectors) {
                args = '-Sx12';
            } else if (JSON.stringify(combined.sparseSelectors) === defaultSelectorsJSON) {
                args = `-Sx${combined.sparseSelectors.length}`;
            } else {
                args = `-S${combined.sparseSelectors.join(',')}`;
            }
            if (typeof combined.precision === 'number') {
                args = `-Zpr${combined.precision} ${args}`;
            }
            if (typeof combined.modelRecipBaseCount === 'number') {
                args = `-Zmd${combined.modelRecipBaseCount} ${args}`;
            }
            if (typeof combined.modelMaxCount === 'number') {
                args = `-Zmc${combined.modelMaxCount} ${args}`;
            }
            if (typeof combined.recipLearningRate === 'number') {
                args = `-Zlr${combined.recipLearningRate} ${args}`;
            }
            if (typeof combined.dynamicModels === 'number') {
                args = `-Zdy${combined.dynamicModels} ${args}`;
            }
            if (typeof combined.numAbbreviations === 'number') {
                args = `-Zab${combined.numAbbreviations} ${args}`;
            }
            return args;
        };

        // FIXME this disables the default SIGINT handling in Windows,
        // so SIGINT would be doing nothing until the next log is reported
        // (when setImmediate is called and signals are actually delivered).
        // the only solution seems to be tick callbacks in compressWithModel etc.
        let stop = false;
        process.once('SIGINT', () => {
            stop = true;
        });

        let result;
        const run = async level => {
            result = await packer.optimize(level, async info => {
                await new Promise(resolve => setImmediate(resolve)); // allow signals to be delivered
                if (verbose >= 0) {
                    console.warn(
                        `(${info.pass}` +
                        (typeof info.passRatio === 'number' ? ` ${(info.passRatio * 100).toFixed(1)}%` : '') +
                        `) ${format(info.current)}:`,
                        info.currentSize, info.bestUpdated ? '<-' : info.currentRejected ? 'x' : '');
                }
                if (stop) {
                    result = info;
                    return false;
                }
            });
            options = { ...options, ...result.best };
        };

        const start = performance.now();
        try {
            if (isFinite(optimize)) {
                await run(optimize);
            } else {
                for (let i = 1; ; ++i) await run(i);
            }
        } catch (e) {
            if (!(e instanceof Error && e.message === 'search aborted')) throw e;
        }
        // do not rely on result.elapsedMsecs which may have been overwritten in -OO
        const elapsedMsecs = performance.now() - start;

        if (verbose >= 0) {
            const ratio = origLength > 0 ? 100 - result.bestSize / origLength * 100 : -Infinity;
            console.warn(
                (stop ? 'search aborted after' : 'search done in') +
                ` ${(elapsedMsecs / 1000).toFixed(1)}s, use \`${format(result.best)}\` to replicate:`,
                result.bestSize,
                `(estimated, ${Math.abs(ratio).toFixed(2)}% ${ratio > 0 ? 'smaller' : 'larger'})`);
        }
    }

    const packed = packer.makeDecoder();
    const output = packed.firstLine + packed.secondLine;
    const compressedLength = packed.estimateLength();
    const ratio = origLength > 0 ? 100 - compressedLength / origLength * 100 : -Infinity;
    if (!optimize && verbose >= 0) {
        console.warn(
            `compressed ${origLength}B into ${compressedLength}B ` +
            `(estimated, ${Math.abs(ratio).toFixed(2)}% ${ratio > 0 ? 'smaller' : 'larger'}).`);
    }
    if (outputPath === '-') {
        console.log(output); // this includes a trailing newline, use the file output to get rid of it
    } else {
        fs.writeFileSync(outputPath, output, { encoding: 'utf-8' });
    }
}

let parsed;
try {
    parsed = await parseArgs(process.argv.slice(2));
} catch (e) {
    if (typeof e === 'string') {
        console.warn(`Error: ${e}`);
        process.exit(1);
    }
    throw e;
}

switch (parsed.command) {
    case 'compress':
        try {
            await compress(parsed);
        } catch (e) {
            // correctly format known classes of errors
            if (e instanceof Error) {
                const m = e.message.match(/^Packer: (.*)$/);
                if (m) {
                    console.warn(`Error: ${m[1]}`);
                    process.exit(1);
                }
            }
            throw e;
        }
        break;

    case 'usage':
        usage(parsed.verbose);
        process.exit(0);

    case 'version':
        console.log(`Roadroller ${VERSION}`);
        process.exit(0);

    default:
        usage(parsed.verbose);
        process.exit(1);
}

